pkg/cloudprovider/provider/anexia/provider.go

```diff
@@ -0,0 +1,346 @@
+	}
+	if status.InstanceID == "" {
+		return nil, cloudprovidererrors.ErrInstanceNotFound
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), anxtypes.GetRequestTimeout)
+	defer cancel()
+
+	info, err := apiClient.VSphere().Info().Get(ctx, status.InstanceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed get machine info: %w", err)
+	}
+
+	return &anexiaInstance{
+		info: &info,
+	}, nil
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (string, string, error) {
+	return "", "", nil
+}
+
+// Create creates a cloud instance according to the given machine
+func (p *provider) Create(machine *v1alpha1.Machine, providerData *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	config, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, "failed to parse MachineSpec: %v", err)
+	}
+
+	apiClient := getClient(config.Token)
+	ctx, cancel := context.WithTimeout(context.Background(), anxtypes.CreateRequestTimeout)
+	defer cancel()
+
+	status, err := getStatus(machine.Status.ProviderStatus)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, "failed to get machine status: %v", err)
+	}
+
+	if status.ProvisioningID == "" {
+		ips, err := apiClient.VSphere().Provisioning().IPs().GetFree(ctx, config.LocationID, config.VlanID)
+		if err != nil {
+			return nil, newError(common.InvalidConfigurationMachineError, "failed to get ip pool: %v", err)
+		}
+		if len(ips) < 1 {
+			return nil, newError(common.InsufficientResourcesMachineError, "no ip address is available for this machine")
+		}
+
+		ipID := ips[0].Identifier
+		networkInterfaces := []anxvm.Network{{
+			NICType: anxtypes.VmxNet3NIC,
+			IPs:     []string{ipID},
+			VLAN:    config.VlanID,
+		}}
+
+		vm := apiClient.VSphere().Provisioning().VM().NewDefinition(
+			config.LocationID,
+			"templates",
+			config.TemplateID,
+			machine.ObjectMeta.Name,
+			config.CPUs,
+			config.Memory,
+			config.DiskSize,
+			networkInterfaces,
+		)
+
+		vm.Script = base64.StdEncoding.EncodeToString(
+			[]byte(fmt.Sprintf("anexia: true\n\n%s", userdata)),
+		)
+
+		sshKey, err := ssh.NewKey()
+		if err != nil {
+			return nil, newError(common.CreateMachineError, "failed to generate ssh key: %v", err)
+		}
+		vm.SSH = sshKey.PublicKey
+
+		provisionResponse, err := apiClient.VSphere().Provisioning().VM().Provision(ctx, vm)
+		if err != nil {
+			return nil, newError(common.CreateMachineError, "instance provisioning failed: %v", err)
+		}
+
+		status.ProvisioningID = provisionResponse.Identifier
+		status.IPAllocationID = ipID
+		if err := updateStatus(machine, status, providerData.Update); err != nil {
+			return nil, newError(common.UpdateMachineError, "machine status update failed: %v", err)
+		}
+	}
+
+	instanceID, err := apiClient.VSphere().Provisioning().Progress().AwaitCompletion(ctx, status.ProvisioningID)
```

Comment:
So if I understand correctly you are blocking the controller here until the progress is if this vm is done, right? You are here using the provisioning id as a progress id, is that correct?

Reply:
Yes, and yes provisioning id is to track progress.